lib/compiler/passes/generate-bytecode.js

```diff
@@ -355,7 +355,27 @@ function generateBytecode(ast) {
     );
   }
 
-  const generate = visitor.build({
+  function wrapGenerators(generators) {
+    if (options.output === "source-and-map") {
+      Object.entries(generators).forEach(([name, generator]) => {
+        generators[name] = function(node, ...args) {
+          return buildSequence(
+            [
+              op.SOURCE_MAP_PUSH,
+              node.location,
+            ],
```

Comment:
Bytecode assumed to be simple byte array. I think you should consider use dedicated fields in `node` to store that information.

You should also update `.d.ts` files

Reply:
Sorry, not quite clear what you mean here? Do you mean I should just do:
```
  [
    op.SOURCE_MAP_PUSH,
    node.location.source,
    node.location.start.line,
    node.location.start.column,
    node.location.end.line,
    node.location.end.column,
  ]
```

Reply:
Now I read code more carefully and understand, that `wrapGenerators` wraps each visitor function into `op.SOURCE_MAP_PUSH...op.SOURCE_MAP_POP` operations.

Well, the canonical way would be to save all locations to the special `locations` array, added here:
https://github.com/peggyjs/peggy/blob/ddb35b4a4ad8df4d3f315fcd79f96352b05d823e/lib/compiler/passes/generate-bytecode.js#L221-L224
and in bytecode use the index into this array (`[op.SOURCE_MAP_PUSH, <index in locations[]>]`).

Then you can add a test that will check that bytecode and constants generated correctly, in https://github.com/peggyjs/peggy/blob/main/test/unit/compiler/passes/generate-bytecode.spec.js. `changeAST` helper accepts an optional third parameter with an `options` object, so you able to test your code:
https://github.com/peggyjs/peggy/blob/ddb35b4a4ad8df4d3f315fcd79f96352b05d823e/test/unit/compiler/passes/helpers.js#L8-L9

Reply:
Makes sense. Will fix.